applications/luci-app-https_dns_proxy/luasrc/model/cbi/https_dns_proxy.lua

```diff
@@ -39,8 +48,129 @@ function uci_add_list(conf, sect, opt, value)
   uci:set(conf, sect, opt, lval)
 end
 
+function servers_restore()
+  local k, v
+	if uci:get("dhcp", "@dnsmasq[0]", "doh_backup_server") then
+		uci:delete("dhcp", "@dnsmasq[0]", "server")
+		for k, v in pairs(uci:get("dhcp", "@dnsmasq[0]", "doh_backup_server")) do
+			uci_del_list("dhcp", "@dnsmasq[0]", "server", v)
+			uci_add_list("dhcp", "@dnsmasq[0]", "server", v)
+		end
+	end
+	uci:commit("dhcp")
+end
+
+function servers_backup()
+  if not uci:get("dhcp", "@dnsmasq[0]", "doh_backup_server") then
+    uci:set("dhcp", "@dnsmasq[0]", "doh_backup_server", uci:get("dhcp", "@dnsmasq[0]", "server"))
+  end
+end
+
+function servers_apply_doh()
+	local n = 0
+	uci:foreach("https_dns_proxy", "https_dns_proxy", function(s)
+		local la_val, lp_val
+		la_val = s[".listen_addr"] or "127.0.0.1"
+		lp_val = s[".listen_port"] or n + 5053
+		uci_del_list("dhcp", "@dnsmasq[0]", "server", tostring(la_val) .. "#" .. tostring(lp_val))
+		uci_add_list("dhcp", "@dnsmasq[0]", "server", tostring(la_val) .. "#" .. tostring(lp_val))
+    n = n + 1
+		uci:save("dhcp")
+	end)
+	uci:commit("dhcp")
+end
+
+function get_provider_name(value)
+  if value:match("dns\.adguard") then
+    return "AdGuard (Standard)"
+  elseif value:match("family\.adguard") then
+    return "AdGuard (Family Protection)"
+  elseif value:match("cleanbrowsing\.org/doh/security") then
+    return "CleanBrowsing (Security Filter)"
+  elseif value:match("cleanbrowsing\.org/doh/family") then
+    return "CleanBrowsing (Family Filter)"
+  elseif value:match("cleanbrowsing\.org/doh/adult") then
+    return "CleanBrowsing (Adult Filter)"
+  elseif value:match("cloudflare") then
+    return "Cloudflare"
+  elseif value:match("gesellschaft\.ch") then
+    return "Digitale Gesellschaft (ch)"
+  elseif value:match("dns\.sb") then
+    return "DNS.SB"
+  elseif value:match("google") then
+    return "Google"
+  elseif value:match("odvr\.nic\.cz") then
+    return "ODVR (nic.cz)"
+  elseif value:match("dns\.quad9") then
+    return "Quad 9 (Recommended)"
+  elseif value:match("dns9\.quad9") then
+    return "Quad 9 (Secured)"
+  elseif value:match("dns10\.quad9") then
+    return "Quad 9 (Unsecured)"
+  elseif value:match("dns11\.quad9") then
+    return "Quad 9 (Secured with ECS Support)"
+  else
+    return "Uknown Provider"
+  end
+end
+
+local tmpfsStatus, tmpfsStatusCode
+local ubusStatus = util.ubus("service", "list", { name = packageName })
+local tmpfsVersion = tostring(util.trim(sys.exec("opkg list-installed " .. packageName .. " | awk '{print $3}'")))
+
+if not tmpfsVersion or tmpfsVersion == "" then
+  tmpfsStatusCode = -1
+  tmpfsVersion = ""
+  tmpfsStatus = packageName .. " " .. translate("is not installed or not found")
```

Comment:
You have already add a dependency to *httpy_dns_proxy* to your Makefile.
https://github.com/openwrt/luci/blob/7707b679e20df9806fbcbc58d12116f9ee9a3850/applications/luci-app-https_dns_proxy/Makefile#L11
So if we install *luci-app-https_dnx_proxy* then the dependency get installed by opkg wo why do we need to check if *https_dns_proxy* is installed?

Reply:
Ah, I see what you mean. Well, the luci app can still be force-installed without the principal package, that's why there's a code in place to handle missing version information from the principal package.

Reply:
From my point of view if I am not an expert of LuCI. Does this really make sense to force an installation of an package if the package system wants to prevent me to install this package.

If so we have a problem in **all** other package that have an dependency to an package. So we have to check this in the code as well. Why query extra the dependency when the package system opkg already offers this feature?

Reply:
> Why query extra the dependency when the package system opkg already offers this feature?

Because I want to display the version of the installed principal service. If opkg returns nothing when I query for the principal service version, what do you suggest luci app do?